examples/training/classification/imagenet/basic_training.py

```diff
@@ -182,49 +202,114 @@ def augment(img, label):
         **eval(FLAGS.model_kwargs),
     )
 
-
 """
-Next, we pick an optimizer. Here we use Adam with a constant learning rate.
-Note that learning rate will decrease over time due to the ReduceLROnPlateau callback.
+Optional LR schedule with cosine decay instead of ReduceLROnPlateau
+TODO: Replace with Core Keras LRWarmup when it's released. This is a temporary solution.
 """
 
 
-optimizer = optimizers.SGD(
-    learning_rate=INITIAL_LEARNING_RATE, momentum=0.9, global_clipnorm=10
+def lr_warmup_cosine_decay(
+    global_step, warmup_steps, hold=0, total_steps=0, start_lr=0.0, target_lr=1e-3
+):
+    # Cosine decay
+    # There is no tf.pi :(
+    learning_rate = (
+        0.5
+        * target_lr
+        * (
+            1
+            + tf.cos(
+                tf.constant(np.pi)
+                * (global_step - warmup_steps - hold)
+                / float(total_steps - warmup_steps - hold)
+            )
+        )
+    )
+    warmup_lr = target_lr * (global_step / warmup_steps)
+
+    if hold > 0:
+        learning_rate = tf.where(
+            global_step > warmup_steps + hold, learning_rate, target_lr
+        )
+
+    learning_rate = tf.where(global_step < warmup_steps, warmup_lr, learning_rate)
+    return learning_rate
+
+
+class WarmUpCosineDecay(keras.optimizers.schedules.LearningRateSchedule):
+    def __init__(self, start_lr, target_lr, warmup_steps, total_steps, hold):
+        super().__init__()
+        self.start_lr = start_lr
+        self.target_lr = target_lr
+        self.warmup_steps = warmup_steps
+        self.total_steps = total_steps
+        self.hold = hold
+
+    def __call__(self, step):
+        lr = lr_warmup_cosine_decay(
+            global_step=step,
+            total_steps=self.total_steps,
+            warmup_steps=self.warmup_steps,
+            start_lr=self.start_lr,
+            target_lr=self.target_lr,
+            hold=self.hold,
+        )
+
+        return tf.where(step > self.total_steps, 0.0, lr, name="learning_rate")
+
+
+# If batched
+total_steps = len(train_ds) * FLAGS.epochs
```

Comment:
The DS isn't loaded at this point so getting its `len` errors.

I am testing the script by using 
`total_steps = (1281167 // BATCH_SIZE) * FLAGS.epochs`

Reply:
How come it's not loaded there yet? Mapping and prefetching is way before that line

Reply:
IIRC the tf.data pipeline is lazy loaded, so it's not loaded until it gets used during training

Reply:
It's not loaded, but it might give us the cardinality

Reply:
This is still not working correctly, and I am testing it using the manual length from above.

I think we can specify NUM_RECORDS and calculate it manually, or we can figure out how to correctly read this from `train_ds`

Reply:
What was the error's name? I can't reproduce it with other datasets (maybe it's only applicable to TFRecords?)
Since it might actually be impossible to get it working, a `NUM_RECORDS` is probably the best option

Reply:
I think it is TFRecords related -- let's just add `NUM_RECORDS`